### PR TITLE
Sketch Linker implementation.

### PR DESCRIPTION
**Pull request by evacchi**
I started looking at the Go implementation, and I think this is VERY ROUGHLY what happens there. In particular, take a look at 
`LinkedModules link()` in `Linker`. 

Essentially the flow is the same as before, except that, when a `Plugin` is created with the `Plugin.Builder`, on `build()` the `Linker` is instantiated and wires together all the `Module.Builder` instances. (Yes, this is very Java). I moved most of the logic from the `Plugin` constructor to this new `Linker` class.

Adding a bunch of folks as reviewer, feel free to add some comments if you have some time. 

Will backfill the tests, hence why this is a Draft. Good thing the single-module version still works, meaning at least I am doing _something_ right.

### CODE REVIEW

File: src/main/java/org/extism/chicory/sdk/ChicoryModule.java [hunk 1]
```diff
@@ -3,23 +3,14 @@
 import com.dylibso.chicory.aot.AotMachine;
 import com.dylibso.chicory.runtime.Module;
 
-class ManifestModuleMapper {
-    private final Manifest manifest;
+public class ChicoryModule {
```
Comment (evacchi) [inline src/main/java/org/extism/chicory/sdk/ChicoryModule.java:6]: Removed this `ManifestModuleMapper`, which has now become the `Linker` and moved some ancillary code inside this `ChicoryModule`

File: src/main/java/org/extism/chicory/sdk/Linker.java [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+
+    /*
+     * Try to find the main module:
+     *  - There is always one main module
+     *  - If a Wasm value has the Name field set to "main" then use that module
+     *  - If there is only one module in the manifest then that is the main module by default
+     *  - Otherwise the last module listed is the main module
+     *
+     */
+    public LinkedModules link() {
```
Comment (chrisdickinson) [inline src/main/java/org/extism/chicory/sdk/Linker.java:38]: I'm not how applicable this will be to the chicory-sdk, but! In the js-sdk, we split up the [manifest resolution](https://github.com/extism/js-sdk/blob/main/src/manifest.ts#L102) and [linking steps](https://github.com/extism/js-sdk/blob/main/src/foreground-plugin.ts#L155). Calling out some points of interest:

- The linker is recursive (so we visit the imports the main module, then the imports of those imports, etc etc.)
- WASIp1 command modules (and their subtree) are instantiated at _every_ point where they're imported.
- In the js-sdk, we allow circular imports of functions, but not of the other export types – JS doesn't give us a straightforward way to generate and inject table/memory/global resources, unfortunately.

Comment (zshipko) [reply src/main/java/org/extism/chicory/sdk/Linker.java:38]: I think the recursive import resolution is a nice way to handle this - that also allows you to skip instantiating modules that aren't actually needed.

Comment (evacchi) [reply src/main/java/org/extism/chicory/sdk/Linker.java:38]: interesting, so Manifest handling is already [handled](https://github.com/extism/chicory-sdk/blob/main/src/main/java/org/extism/chicory/sdk/ManifestWasm.java) [separately](https://github.com/extism/chicory-sdk/blob/main/src/main/java/org/extism/chicory/sdk/Manifest.java#L28); 

There are also two separate phases in the linker; i.e. the Linker here only resolves the modules (sure, not recursively) but won't instantiate. Instead, it returns a `LinkedModule` and then `LinkedModule#toPlugin` instantiates all `Module`(`.Builder`s) at once. 

Indeed, I am not cross-linking imports though 😬  I am quite sure that at least part of this logic belongs to Chicory itself 🤔

Comment (evacchi) [reply src/main/java/org/extism/chicory/sdk/Linker.java]: followed up to this original work rebasing the work on the Store that has been introduced with https://github.com/dylibso/chicory/pull/495

this is still unfinished, at the bare minimum it lacks proper testing; however we can already see that the Kernel is now instantiated as a consequence of the requirement in the `main` module, instead of being provided regardless (not necessarily the desired behavior, but at least it shows that it works for this simple example :D)

File: src/main/java/org/extism/chicory/sdk/ChicoryModule.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+import com.dylibso.chicory.aot.AotMachine;
+import com.dylibso.chicory.runtime.Instance;
+import com.dylibso.chicory.wasm.Module;
+import com.dylibso.chicory.wasm.Parser;
+
+import java.nio.file.Path;
+
+public class ChicoryModule {
+
+    static final boolean IS_NATIVE_IMAGE_AOT = Boolean.getBoolean("com.oracle.graalvm.isaot");
```
Comment (evacchi) [inline src/main/java/org/extism/chicory/sdk/ChicoryModule.java:12]: this is not strictly related to this PR, but it's necessary to compile this under graalvm native image. This flag is true when the native image builder it's running. We want to make sure that the code path below is pruned at build-time, because the AoT in its current state is not compatible with the way native image works.

File: pom.xml [hunk 4]
```diff
@@ -18,23 +18,25 @@
     <maven.compiler.source>11</maven.compiler.source>
     <maven.compiler.target>11</maven.compiler.target>
     <maven.compiler.release>11</maven.compiler.release>
+
+    <chicory.version>999-SNAPSHOT</chicory.version>
```
Comment (evacchi) [inline pom.xml:22]: I suppose we might want to wait for a release before merging

**Comment by evacchi** [comment]
I can finally rebase this on https://github.com/dylibso/chicory/releases/tag/1.0.0-M1 😅
